src/core/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -1,64 +1,105 @@
+#include <core/platform/f4xx/stm32f4xx.hpp>
+#include <core/utils/log.hpp>
+
+namespace {
+/// This is made as a global variable so that it is accessible in the interrupt.
+DMA_HandleTypeDef* dmaHandle[3];
+ADC_HandleTypeDef* adcHandle[3];
+} // namespace
+
+/**
+ * @brief This function handles DMA2 stream0 global interrupt. (For ADC 1)
+ */
+extern "C" void DMA2_Stream0_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[0]);
+    HAL_ADC_IRQHandler(adcHandle[0]);
+}
+
+/**
+ * @brief This function handles DMA2 stream2 global interrupt. (For ADC 2)
+ */
+extern "C" void DMA2_Stream2_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[1]);
+    HAL_ADC_IRQHandler(adcHandle[1]);
+}
+
+/**
+ * @brief This function handles DMA2 stream1 global interrupt. (For ADC 3)
+ */
+extern "C" void DMA2_Stream1_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[2]);
+    HAL_ADC_IRQHandler(adcHandle[2]);
+}
 
 namespace core::io {
+#define ADC1_SLOT 0
+#define ADC2_SLOT 1
+#define ADC3_SLOT 2
+
+#define ADC1SHIFT 5
+#define ADC2SHIFT 6
+#define ADC3SHIFT 7
 
-// Init static member variables
-ADC_HandleTypeDef ADCf4xx::halADC = {0};
-Pin ADCf4xx::channels[MAX_CHANNELS];
-uint16_t ADCf4xx::buffer[MAX_CHANNELS];
-DMA_HandleTypeDef ADCf4xx::halDMA = {0};
+// Combines the channel memory value with the ADC peripherals it supports into one uint32_t
+#define CHANNEL_SET(adc1, adc2, adc3, ch) (ch | (adc1 << ADC1SHIFT) | (adc2 << ADC2SHIFT) | (adc3 << ADC3SHIFT))
 
-ADCf4xx::ADCf4xx(Pin pin) : ADC(pin) {
-    // Flag representing if the ADC has been configured yet
-    static bool halADCisInit = false;
+ADCf4xx::ADC_State_t ADCf4xx::adcArray[3];
+bool ADCf4xx::timerInit = false;
 
-    // "Rank" represents the order in which the channels are added
-    // Also represents the total number of added channels
-    static uint8_t rank = 1;
+ADCf4xx::ADCf4xx(Pin pin, ADCPeriph adcPeriph) : ADC(pin, adcPeriph) {
+    // Get adc state being updated
+    uint8_t adcNum                 = getADCNum();
+    ADCf4xx::ADC_State_t* adcState = &adcArray[adcNum];
 
-    // Maximum number of ADC channels have already been added
-    if (rank == MAX_CHANNELS) {
+    if (adcState->rank == MAX_CHANNELS) {
+        log::LOGGER.log(log::Logger::LogLevel::WARNING, "ADC %d ALREADY HAS MAX PINS!!", (adcNum + 1));
         return;
     }
 
-    // Initialization of the HAL ADC should only take place once since there is
-    // only one ADC device which has multiple channels supported.
-    if (halADCisInit) {
-        HAL_ADC_Stop_DMA(&halADC);
-        HAL_DMA_DeInit(&halDMA);
+    if (timerInit) {
+        HAL_TIM_Base_DeInit(&htim8); // Stop Timer8 (Trigger Source For ADC's)
+        timerInit = false;
+    }
+
+    if (adcState->isADCInit) {
+        HAL_ADC_Stop_DMA(&adcState->halADC);
+        HAL_DMA_DeInit(&adcState->halDMA);
     } else {
         __HAL_RCC_DMA2_CLK_ENABLE();
-        halADCisInit = true;
+        adcState->isADCInit = true;
     }
```

Comment:
I'm kinda confused by this comment. The isADCInit is unique to each adc instance, so it will be false for the first time that the specific adc peripheral is started (1, 2, or 3), and then any channels added to that specific adc after that it will do the true part of the if statement, which stops the ADC & associated DMA, and then it restarts the ADC and DMA in the constructor later.

If your talking about the __DMA2_CLK_ENABLE(), that just sets a bit to 1, so you can call it as many times as we want, and it will be fine...

Reply:
For some reason, I thought `isADCInit` was getting set to `false` somewhere. I think this should be fine